core/sr-primitives/src/sr_arithmetic.rs

```diff
@@ -0,0 +1,1013 @@
+				let p: $upper_type = self.0 as $upper_type * self.0 as $upper_type;
+				let q: $upper_type = $max as $upper_type * $max as $upper_type;
+				Self::from_rational_approximation(p, q)
+			}
+
+			/// Converts a fraction into `Permill`.
+			#[cfg(feature = "std")]
+			pub fn from_fraction(x: f64) -> Self { Self((x * ($max as f64)) as $type) }
+
+			/// Approximate the fraction `p/q` into a per-thing fraction. This will never overflow.
+			///
+			/// The computation of this approximation is performed in the generic type `N`. Given
+			/// `M` as the data type that can hold the maximum value of this per-thing (e.g. u32 for
+			/// perbill), this can only work if `N == M` or `N: From<M> + TryInto<M>`.
+			pub fn from_rational_approximation<N>(p: N, q: N) -> Self
+				where N: Clone + Ord + From<$type> + TryInto<$type> + ops::Div<N, Output=N>
+			{
+			// TODO: double check with @thiolliere the new bound. If your fraction type has a max
+			// that fits in u64, then you cannot build a rational from it that is `2u8 / 4u8` So
+			// using `SimpleArithmetic` here was okay as long as we don't add a new fraction type
+			// which is embedded in a u64. With perquintill we have to change this.
+				// q cannot be zero.
+				let q = q.max((1 as $type).into());
+				// p should not be bigger than q.
+				let p = p.min(q.clone());
+
+				let factor = (q.clone() / $max.into()).max((1 as $type).into());
+
+				// q cannot overflow: (q / (q/$max)) < $max. p < q hence p also cannot overflow.
+				let p_reduce: $type = (p / factor.clone()).try_into().unwrap_or_else(|_| panic!());
+				let q_reduce: $type = (q / factor.clone()).try_into().unwrap_or_else(|_| panic!());
+				// `p_reduced` and `q_reduced` are withing $type. Mul by another $max will always
+				// fit in $upper_type. This is guaranteed by the macro tests.
+				let part =
+					p_reduce as $upper_type
+					* ($max as $upper_type)
+					/ q_reduce as $upper_type;
+
+				$name(part as $type)
+			}
+		}
+
+		impl Saturating for $name {
+			fn saturating_add(self, rhs: Self) -> Self {
+				Self::from_parts(self.0 + rhs.0)
+			}
+			fn saturating_sub(self, rhs: Self) -> Self {
+				Self::from_parts(self.0.saturating_sub(rhs.0))
+			}
+			fn saturating_mul(self, rhs: Self) -> Self {
+				let a = self.0 as $upper_type;
+				let b = rhs.0 as $upper_type;
+				let m = $max as $upper_type;
+				let parts = a * b / m;
+				// This will always fit into $type.
+				Self::from_parts(parts as $type)
+			}
+		}
+
+		impl ops::Div for $name {
+			type Output = Self;
+
+			fn div(self, rhs: Self) -> Self::Output {
+				let p = self.0;
+				let q = rhs.0;
+				Self::from_rational_approximation(p, q)
+			}
+		}
+
+		/// Overflow-prune multiplication.
+		///
+		/// tailored to be used with a balance type.
+		impl<N> ops::Mul<N> for $name
+		where
+			N: Clone + From<$type> + UniqueSaturatedInto<$type> + ops::Rem<N, Output=N>
+				+ ops::Div<N, Output=N> + ops::Mul<N, Output=N> + ops::Add<N, Output=N>,
```

Comment:
Yeah, changing `From<u32> + UniqueSaturatedInto<u32>` to `From<$type> + UniqueSaturatedInto<$type>`  actually only makes a difference for that. But I think it is mandatiry: 

correct me if I am wrong but afair you added this implementation with the mindset that:
2- it operates in type `N` which the user provides.
1- it does not work for all `N` types but if does, then it does not overflow. 

Before, both perthings were smaller than u32 and `From<u32>` was enough. Perquintill needs u64 and it is natural that if you want to, for example, multiply `10` by `Perquintill::from_percent(50)`, you need to pass `10` as `10u64` (or `10u128` in this case), hence delcaring `N` as `u64` which makes all operations safe.


> if $max.try_into<N>() is err that means that N fits into $type so we could just make use of upper type no ?
> maybe we should keep as it is though and user will just require a new bound on its type he wants to multiply with perquintill

But I can't get what you mean here, let me know if I am missing something or my argument above is incorrect.

Reply:
no everything you say is correct just maybe we can change a bit the behavior so that PerQuintill works with SimpleArithmetic.

>> if $max.try_into() is err that means that N fits into $type so we could just make use of upper type no ?

This means: for PerQuintill the change can be made this way: use the bound `TryInto<u64>` instead of `From<u64>` which is not on SimpleArithmetic.
```rust
				let maximum: N = $max.try_into();
```
Here if maximum is err that means that N is less u64 thus `b: N` can be converted to u64 without any issue. if maximum is not err then fine same as now.

>> maybe we should keep as it is though and user will just require a new bound on its type he wants to multiply with perquintill

I think my above proposal is too complex and probably doesn't worth it. the user can add an additional bound on its type if he wants to.